File: solutions/opus-1/opus/java/dijkstra.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        DataInputStream in = new DataInputStream(new BufferedInputStream(System.in, 1 << 16));
        int n = (int) next(in), m = (int) next(in);
        int[] head = new int[Math.max(n, 1)];
        Arrays.fill(head, -1);
        int[] nxt = new int[2 * m], to = new int[2 * m];
        long[] wt = new long[2 * m];
        int e = 0;
        for (int i = 0; i < m; i++) {
            int u = (int) next(in), v = (int) next(in);
            long w = next(in);
            to[e] = v; wt[e] = w; nxt[e] = head[u]; head[u] = e++;
            to[e] = u; wt[e] = w; nxt[e] = head[v]; head[v] = e++;
        }
        int s = (int) next(in), t = (int) next(in);
        if (s == t) { System.out.println(0); return; }
        long[] dist = new long[n];
        Arrays.fill(dist, Long.MAX_VALUE);
        dist[s] = 0;
        PriorityQueue<long[]> pq = new PriorityQueue<>((a, b) -> Long.compare(a[0], b[0]));
        pq.add(new long[]{0, s});
        while (!pq.isEmpty()) {
            long[] cur = pq.poll();
            int u = (int) cur[1];
            if (cur[0] > dist[u]) continue;
            if (u == t) break;
            for (int k = head[u]; k != -1; k = nxt[k]) {
                long nd = cur[0] + wt[k];
                if (nd < dist[to[k]]) {
                    dist[to[k]] = nd;
                    pq.add(new long[]{nd, to[k]});
                }
            }
        }
        System.out.println(dist[t] == Long.MAX_VALUE ? -1 : dist[t]);
    }

    private static long next(DataInputStream in) throws IOException {
        int c = in.read();
        while (c != '-' && (c < '0' || c > '9')) c = in.read();
        boolean neg = c == '-';
        if (neg) c = in.read();
        long r = 0;
        while (c >= '0' && c <= '9') { r = r * 10 + (c - '0'); c = in.read(); }
        return neg ? -r : r;
    }
}

File: solutions/opus-1/opus/javascript/dijkstra.js
const data = require('fs').readFileSync(0, 'utf8');
let pos = 0;
function next() {
  while (pos < data.length) {
    const c = data.charCodeAt(pos);
    if (c >= 48 && c <= 57) break;
    pos++;
  }
  let x = 0;
  while (pos < data.length) {
    const c = data.charCodeAt(pos);
    if (c < 48 || c > 57) break;
    x = x * 10 + (c - 48);
    pos++;
  }
  return x;
}

const n = next(), m = next();
const head = new Int32Array(n).fill(-1);
const nxt = new Int32Array(2 * m);
const to = new Int32Array(2 * m);
const wt = new Float64Array(2 * m);
let ec = 0;
for (let i = 0; i < m; i++) {
  const u = next(), v = next(), w = next();
  to[ec] = v; wt[ec] = w; nxt[ec] = head[u]; head[u] = ec++;
  to[ec] = u; wt[ec] = w; nxt[ec] = head[v]; head[v] = ec++;
}
const s = next(), t = next();

if (s === t) {
  console.log('0');
} else {
  const dist = new Float64Array(n).fill(Infinity);
  const hd = [], hn = [];
  const push = (d, v) => {
    let i = hd.length;
    hd.push(d); hn.push(v);
    while (i > 0) {
      const p = (i - 1) >> 1;
      if (hd[p] <= d) break;
      hd[i] = hd[p]; hn[i] = hn[p];
      i = p;
    }
    hd[i] = d; hn[i] = v;
  };
  const pop = () => {
    const ld = hd.pop(), ln = hn.pop();
    const size = hd.length;
    if (size === 0) return;
    let i = 0;
    while (true) {
      let c = 2 * i + 1;
      if (c >= size) break;
      if (c + 1 < size && hd[c + 1] < hd[c]) c++;
      if (hd[c] >= ld) break;
      hd[i] = hd[c]; hn[i] = hn[c];
      i = c;
    }
    hd[i] = ld; hn[i] = ln;
  };
  dist[s] = 0;
  push(0, s);
  let ans = -1;
  while (hd.length > 0) {
    const d = hd[0], u = hn[0];
    pop();
    if (d > dist[u]) continue;
    if (u === t) { ans = d; break; }
    for (let e = head[u]; e !== -1; e = nxt[e]) {
      const v = to[e], nd = d + wt[e];
      if (nd < dist[v]) { dist[v] = nd; push(nd, v); }
    }
  }
  console.log(String(ans));
}
